// Representation of game state.

message PlayerInfo {
  enum Life {
    ALIVE = 0;
    DEAD = 1;
    ZOMBIE = 2;
    ZOMBIE_DEAD = 3;
  }

  required uint32 player_id = 1;
  required string name = 2;
  optional uint64 start_tick = 3;
  optional Life alive = 4;
  optional int32 score = 5;
  repeated Block.Type inventory = 6;
  repeated Block power_up = 7;
}

message Coordinate {
  required int32 x = 1;
  required int32 y = 2;
}

message Block {
  enum Type {
    EMPTY = 0;
    PLAYER_HEAD = 1;
    PLAYER_TAIL = 2;
    WALL = 3;
    AMMO = 4;
    ROCKET = 5;
    MINE = 6;
    ROCK = 7;
    BROKEN_ROCK = 8;
    TREE = 9;

    // power-ups
    STAY_STILL = 10;
    FAST = 11;
    TELEPORT = 12;
    INVINCIBLE = 13;

    // special weapons
    NUKE = 14;
  }

  required Type type = 1;
  required Coordinate pos = 2;
  optional Coordinate direction = 3;
  optional uint64 last_viable_tick = 4;
  optional uint32 player_id = 5;
  optional bool move = 6 [default = true];  // If false, do not move this block.
}

message Stage {
  enum Id {
    COLLECT_PLAYERS = 1;
    ROUND = 2;
    ROUND_END = 3;
    GAME_OVER = 4;
  }
}

message Mode {
  enum Id {
    BATTLE = 1;
    CLEAR_MINES = 2;
  }
}
